File: hackit/wireless/plugins/lua/scripts/client_hunt.lua
local iface = arg[1]
local bssid = arg[2] or ""
local timeout = tonumber(arg[3]) or 60

local function json_escape(s)
  return '"' .. tostring(s):gsub('\\', '\\\\'):gsub('"', '\\"'):gsub('\n', '\\n'):gsub('\r', '\\r') .. '"'
end

local function telemetry(event, data)
  local ts = tostring(os.time())
  io.write('{"event":' .. json_escape(event) .. ',"iface":' .. json_escape(iface) .. ',"bssid":' .. json_escape(bssid) .. ',"ssid":"","channel":"","data":' .. data .. ',"timestamp":' .. json_escape(ts) .. '}\n')
  io.flush()
end

local function check_monitor_mode()
  local f = io.popen("iw dev " .. iface .. " info 2>/dev/null")
  if not f then return false end
  local out = f:read("*a")
  f:close()
  return out:find("type monitor") ~= nil
end

local clients = {}
local signal_history = {}

local function normalize_mac(mac)
  if not mac then return nil end
  return mac:lower():gsub("-", ":")
end

local function parse_probe_request(line)
  local sa = line:match("SA:([%x:]+)")
  local ssid = line:match("Probe Request%s+%([^)]+%)%s+(.+)") or line:match("SSID:[%s]+([^\n]+)")
  if not sa then
    sa = line:match("([%x:][%x:]:[%x:][%x:]:[%x:][%x:]:[%x:][%x:]:[%x:][%x:]:[%x:][%x:])")
  end
  if sa then
    sa = normalize_mac(sa)
    if not clients[sa] then
      clients[sa] = {first_seen = os.time(), probes = {}, signals = {}}
    end
    if ssid and ssid ~= "" then
      ssid = ssid:gsub("^%s*(.-)%s*$", "%1")
      clients[sa].probes[ssid] = (clients[sa].probes[ssid] or 0) + 1
    end
  end
end

local function parse_data_frame(line)
  local src = line:match("SA:([%x:]+)")
  local dst = line:match("DA:([%x:]+)")
  local sig = line:match("signal:([%-%d]+)")
  if not src then
    src = line:match("([%x:][%x:]:[%x:][%x:]:[%x:][%x:]:[%x:][%x:]:[%x:][%x:]:[%x:][%x:])")
  end
  if src then
    src = normalize_mac(src)
    if not clients[src] then
      clients[src] = {first_seen = os.time(), probes = {}, signals = {}}
    end
    if sig then
      local s = tonumber(sig)
      table.insert(clients[src].signals, s)
      if not signal_history[src] then signal_history[src] = {} end
      table.insert(signal_history[src], {time = os.time(), signal = s})
      if #signal_history[src] > 100 then table.remove(signal_history[src], 1) end
    end
    if bssid and #bssid > 0 and dst then
      if normalize_mac(dst) == normalize_mac(bssid) then
        telemetry("client_active", '{"mac":' .. json_escape(src) .. ',"destination":' .. json_escape(dst) .. ',"signal":' .. json_escape(sig or "0") .. '}')
      end
    end
  end
end

local function parse_tshark_output(line)
  if line:find("Probe Request") or line:find("Probe") then
    parse_probe_request(line)
  else
    parse_data_frame(line)
  end
end

local function run_sniffer()
  local cmd = "tshark -i " .. iface .. " -Y 'wlan.fc.type == 0 and (wlan.fc.type_subtype == 4 or wlan.fc.type_subtype == 8 or wlan.fc.type_subtype == 32)' -T fields -e frame.time_relative -e wlan.sa -e wlan.da -e radiotap.dbm_antsignal -e wlan.ssid 2>/dev/null"
  if bssid and #bssid > 0 then
    cmd = "tshark -i " .. iface .. " -Y '(wlan.fc.type == 0) and (wlan.addr1==" .. bssid .. " or wlan.addr2==" .. bssid .. " or wlan.addr3==" .. bssid .. ")' -T fields -e frame.time_relative -e wlan.sa -e wlan.da -e radiotap.dbm_antsignal 2>/dev/null"
  end
  telemetry("sniff_start", '{"cmd":' .. json_escape(cmd) .. ',"timeout":' .. tostring(timeout) .. '}')
  local f = io.popen(cmd, "r")
  if not f then
    telemetry("sniff_fail", '{"error":"Failed to start tshark"}')
    return
  end
  local start = os.time()
  while os.time() - start < timeout do
    local line = f:read("*l")
    if not line then break end
    if line and #line > 5 then
      parse_tshark_output(line)
    end
    if os.time() - start > timeout then break end
  end
  f:close()
end

local function report_clients()
  local count = 0
  for mac, info in pairs(clients) do
    count = count + 1
    local avg_sig = 0
    if #info.signals > 0 then
      for _, s in ipairs(info.signals) do avg_sig = avg_sig + s end
      avg_sig = avg_sig / #info.signals
    end
    local probe_list = ""
    for ssid, cnt in pairs(info.probes) do
      probe_list = probe_list .. ssid .. "(" .. cnt .. ") "
    end
    local elapsed = os.time() - info.first_seen
    telemetry("client_report", '{"mac":' .. json_escape(mac) .. ',"probes":' .. json_escape(probe_list) .. ',"avg_signal":' .. tostring(avg_sig) .. ',"signals_count":' .. tostring(#info.signals) .. ',"first_seen_sec":' .. tostring(elapsed) .. ',"active":' .. tostring(elapsed < 30) .. '}')
  end
  telemetry("client_summary", '{"total_clients":' .. tostring(count) .. '}')
end

if not iface then
  telemetry("error", '{"error":"Usage: client_hunt.lua <interface> [bssid] [timeout]"}')
  os.exit(1)
end

if not check_monitor_mode() then
  telemetry("error", '{"error":"Monitor mode required on ' .. iface .. '"}')
  os.exit(1)
end

local ok, err = pcall(function()
  run_sniffer()
  report_clients()
end)

if not ok then
  telemetry("error", '{"error":' .. json_escape(tostring(err)) .. '}')
  os.exit(1)
end

telemetry("complete", '{"bssid":' .. json_escape(bssid) .. '}')

File: hackit/wireless/plugins/ruby/scripts/client_hunt.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'time'
require 'socket'

def emit(event, iface, bssid, ssid, channel, data)
  puts JSON.generate(event: event, iface: iface, bssid: bssid, ssid: ssid, channel: channel, data: data, timestamp: Time.now.iso8601)
  $stdout.flush
end

ETH_P_ALL = [0x0003].pack('n').unpack1('S')

def if_nametoindex(name)
  Socket.getifaddrs.select { |a| a.name == name }.first&.ifindex
end

def mac_addr(bytes)
  bytes.unpack('H2 H2 H2 H2 H2 H2').map { |b| b.upcase }.join(':')
end

def parse_probe_request(frame, data)
  return nil if frame.size < 24
  fc = frame[0..1].unpack1('v')
  frame_type = (fc >> 2) & 0x3
  frame_subtype = (fc >> 4) & 0xF
  return nil unless frame_type == 0 && frame_subtype == 4

  src = mac_addr(frame[10..15])
  dst = mac_addr(frame[4..9])
  bssid_field = mac_addr(frame[16..21])

  pos = 24
  ssid = nil
  rates = nil
  while pos < frame.size - 2
    tag = frame[pos].ord
    len = frame[pos + 1].ord
    break if pos + 2 + len > frame.size
    val = frame[pos + 2, len]
    # rubocop:disable Style/IfUnlessModifier
    if tag == 0 && len > 0
      ssid = val
    end
    if tag == 1 && len > 0
      rates = val.unpack('C*').map { |r| (r & 0x7F) / 2.0 }
    end
    # rubocop:enable Style/IfUnlessModifier
    pos += 2 + len
  end

  { src: src, dst: dst, bssid_from_frame: bssid_field, ssid: ssid, rates: rates, frame_size: frame.size }
end

def parse_data_frame(frame)
  return nil if frame.size < 24
  fc = frame[0..1].unpack1('v')
  frame_type = (fc >> 2) & 0x3
  frame_subtype = (fc >> 4) & 0xF
  return nil unless frame_type == 2

  addr1 = mac_addr(frame[4..9])
  addr2 = mac_addr(frame[10..15])
  addr3 = mac_addr(frame[16..21])
  to_ds = (fc >> 8) & 0x1
  from_ds = (fc >> 9) & 0x1

  bssid = if to_ds == 0 && from_ds == 0
            addr3
          elsif to_ds == 1 && from_ds == 0
            addr1
          elsif to_ds == 0 && from_ds == 1
            addr2
          else
            addr3
          end

  { addr1: addr1, addr2: addr2, addr3: addr3, bssid: bssid, to_ds: to_ds, from_ds: from_ds }
end

begin
  iface = ARGV[0]
  target_bssid = ARGV[1]
  timeout = (ARGV[2] || '30').to_i

  raise 'interface required' unless iface
  ifidx = if_nametoindex(iface)
  raise "interface #{iface} not found" unless ifidx

  emit('client_hunt_start', iface, target_bssid || '', '', 0, { timeout: timeout })

  sock = Socket.new(Socket::PF_PACKET, Socket::SOCK_RAW, ETH_P_ALL)
  sock.setsockopt(Socket::SOL_SOCKET, Socket::SO_RCVBUF, 2_097_152)

  clients = {}
  start_time = Time.now

  while Time.now - start_time < timeout
    raw = sock.recvfrom(8192)[0]
    rt_len = raw[2..3].unpack1('v') rescue 0
    next if rt_len < 4 || rt_len >= raw.size
    frame = raw[rt_len..]
    next if frame.nil? || frame.size < 24

    probe = parse_probe_request(frame, raw)
    if probe
      unless clients[probe[:src]]
        clients[probe[:src]] = { probes: [], data_frames: [], first_seen: Time.now.iso8601, last_seen: Time.now.iso8601 }
      end
      clients[probe[:src]][:probes] << probe
      clients[probe[:src]][:last_seen] = Time.now.iso8601
      clients[probe[:src]][:bssid_probed] = probe[:bssid_from_frame] if probe[:bssid_from_frame] && probe[:bssid_from_frame] != 'FF:FF:FF:FF:FF:FF'

      if !target_bssid || probe[:bssid_from_frame] == target_bssid || probe[:bssid_from_frame] == 'FF:FF:FF:FF:FF:FF'
        emit('probe_request', iface, probe[:bssid_from_frame], probe[:ssid] || '', 0, {
          client: probe[:src], ssid_probed: probe[:ssid],
          rates: probe[:rates], frame_size: probe[:frame_size],
          clients_count: clients.size
        })
      end
    end

    data = parse_data_frame(frame)
    if data
      unless clients[data[:addr2]]
        clients[data[:addr2]] = { probes: [], data_frames: [], first_seen: Time.now.iso8601, last_seen: Time.now.iso8601 }
      end
      clients[data[:addr2]][:data_frames] << data
      clients[data[:addr2]][:last_seen] = Time.now.iso8601

      if !target_bssid || data[:bssid] == target_bssid
        emit('data_frame', iface, data[:bssid], '', 0, {
          client: data[:addr2], destination: data[:addr1],
          to_ds: data[:to_ds], from_ds: data[:from_ds], bssid_from_data: data[:bssid],
          clients_count: clients.size
        })
      end
    end
  end

  sock.close

  emit('client_hunt_summary', iface, target_bssid || '', '', 0, {
    clients_found: clients.size,
    clients: clients.map { |mac, info| { mac: mac, probes: info[:probes].size, data_frames: info[:data_frames].size, first_seen: info[:first_seen], last_seen: info[:last_seen] } },
    timeout: timeout
  })

  emit('client_hunt_complete', iface, target_bssid || '', '', 0, { clients_found: clients.size })

rescue Interrupt
  emit('client_hunt_interrupted', iface || '', target_bssid || '', '', 0, {})
rescue => e
  $stderr.puts JSON.generate(event: 'error', data: { message: e.message, backtrace: e.backtrace&.first }, timestamp: Time.now.iso8601)
  exit 1
end
